### ISSUE TITLE
Custom sun elevation angle -event [Feature request]

### ISSUE THREAD

**Issue by Timo-Fi** [forrestguice/SuntimesWidget#598]
Would be helpful at least for sunbathing purposes. One degree resolution would be well enough. Only one item would be needed in the event list, "custom sun angle". But needs additional value element with the qualifier option somewhere, or only the value element if "+" means ascending sun and "-" means descending sun.

A sunbather could then set an alarm to trigger at e.g. at 45° rising sun (usable UV level starts), with e.g. a 20 min before offset (driving time).

**Comment by forrestguice** [comment]
What I imagine is the app allowing custom angles - from `Settings->User Interface` there could be a list of custom events (0-9 for simplicity) that could be assigned some arbitrary angle, and then checked/unchecked. These could then be displayed in the main table, or displayed by a widget, or assigned to an alarm. In this case the +/- would indicate above/below the horizon (and the app would automatically offer a rising and setting entry when requesting a selection).

This would be separate from a "degree offset" feature, which would allow a before/after offset based on degrees. It seems the offset feature is just extra at that point - having custom angles is essentially the same thing. Also, future versions (further extended), might add UI to the alarm picker to allow selecting arbitrary angles (vs using a custom preset).

Since `v0.14.0` the app is in a good position to support this, but these would be proposals for `v0.15.0` (or possibly later). To set expectations, these releases are usually many months apart - leaves will probably fall off the trees first. If its sooner, it probably means I'm spending too much time inside. ;)

**Comment by Timo-Fi** [comment]
Hi, this is not an important feature, please do not hurry this at all. It will be helpful, and certainly will add yet an other wow feature.  And yes, if you do the custom degree offset, it will make this custom elevation degree completely unnecessary, because a fixed elevation° can then be defined as sunrise+offset°  or sunset-offset°. ''

Thank you!

**Comment by al-caveman** [comment]
I came here to submit this feature request, and found it here.

Basically, I think, it would be great if we could add custom sun times beyond the ones listed (e.g. twilight, sunrise, moon, sunset), so that we could add newer ones, such as:

- Suntime midnight; the middle time between sunset and sunrise (e.g. the opposite of noon).  This is different than 00:00AM as the suntime midnight could be around 1AM in winter (depending on the country of course).

**Comment by Timo-Fi** [comment]
Hello Toraboran Caveman, 

I take it you're suggesting the ability to add new/custom events, so expanding the events list by custom items (event_name, custom_angle). That would be very nice for me also.  It could perhaps be easy to implement, and the icon could be just a simple "C" for custom.  

Aha, but the solar midnight needs calculation, rather than a user defined angle.  On the other hand, solar midnight happens at -12h or +12h from the solar noon so that is already available!

Greetings,
Timo

**Comment by al-caveman** [comment]
> ... On the other hand, solar midnight happens at -12h or +12h from the solar noon so that is already available!

Oh neat. Thanks.  How about this solar time:

> The time that starts at the moment when a perpendicular stick's shadow equals _twice_ stick's height, plus the length of stick's shadow at noon (Zenith)

Is there any easy way to achieve this?

This is yet another afternoon sub-time, that happens before the evening golden hour.  It is commonly used today at least in Northern Africa, Middle East and Central Asia.

**Comment by Timo-Fi** [comment]
Hi, my understanding is that, currently, any event that can be bound by a time difference to one of those events found in Suntimes can be defined as an alarm/notification. 

To me it seems that something is probably missing from the definition you posted, since the length of stick's shadow at solar noon depends on the day of the year (for one thing). If the following is added to the end of it: "at the summer solstice", then it would be a defined altitude/elevation of the sun (at chosen topographic elevation level). Even then, such alarm/notification is not possible currently.

**Comment by forrestguice** [comment]
An update on this feature - It will definitely be in `v0.15.0` (probably sometime before the new year). That code is in the [events](https://github.com/forrestguice/SuntimesWidget/tree/events) branch.

I will consider adding 'solar midnight' to the list of fields. I don't see why not. I'm not sure I fully understand its 'use case' though.

> The time that starts at the moment when the shadow length of a stick, that is perpendicular on the ground, equals twice as much as stick's height, plus the length of stick's shadow at noon (Zenith).

@al-caveman I consider topics like this really interesting. I'd like to read more about it. Does this definition go by some name or phrase that I can research? I think it's probably out of scope for Suntimes itself, but I can easily imagine it taking the form of an addon.

> Thanks a lot for making this app. It's the best by far (I was surprised it is not in Google Store. Any reason?).

Thank you for the praise.
As for Google Store, I have my reasons. I'm able to release apps there if I choose, and I was pretty enthusiastic about that when I first registered. I'm less enthusiastic now. I voluntarily unpublished my apps several years ago, rather than update them to meet ever changing guidelines/requirements (and risk some violation). 

Suntimes was written as an F-Droid exclusive. I won't rule out the possibility of publishing to Google Store, but as things are now its not even possible; the app targets api25 (Android 7), but Google requires the latest (now at 31+). Its been [locked at 25](https://github.com/forrestguice/SuntimesWidget/issues/122) since v26, when the support libraries increased the minSdk to 14 (dropping support for "ancient" devices). I want to continue supporting "old devices" (however defined) because I don't like the [environmental](https://www.google.com/maps/@34.6084891,-113.2623873,14347m/data=!3m1!1e3) [impact](https://www.hcn.org/articles/indigenous-affairs-justice-at-oak-flat-courts-and-politicians-fail-tribes) of constantly producing "new devices" (planned obsolescence). Its been a few years, and I'm now actually planning to bump the minSdk to 14, but I've discovered benefits to lagging behind Google requirements.. The work required (technical debt) discourages those who would republish the app with undisclosed modifications (i.e. malware), or otherwise ignore the terms of the license.

**Comment by al-caveman** [comment]
@forrestguice  - Thanks.

This time is called [Asr](https://en.wikipedia.org/wiki/Asr_prayer).  It is a well-known ancient time that was also eventually used a prayer time sometime in history (~1400 years ago).  Now that 1400 years have passed, it is easier to search for "Asr prayer time".  However, as a time, it predates the moment when it was a prayer time, and many people still use it beyond the context of this prayer.

As for Google Play, I don't know much about its policies.  Are you saying that Google requires that, if an app supports, say, API 14+, Google rejects it, despite the fact that API 14+ also supports 31+?  I.e. Google forces developers to artificially forbid too-old devices from being able to get an app, even if newer devices are still supported?

---

@Timo-Fi 

> To me it seems that something is probably missing from the definition you posted, since the length of stick's shadow at solar noon depends on the day of the year (for one thing).

What you describe is surely ambiguous, because shadow's length depends on the year day (as you have rightfully said).  However, that's not what I said.

You missed this criteria:  this length doubling metric takes into account a reference shadow at noon time.  Specifically, whatever is the shadow length at noon is used as a normalisation value to deduct from shadow lengths.

Basically, it's `shadow_length - offset == stick_height`, where `offset` is the shadow length of the same stick at the noon of the same day.

Does this address the issue?  Or am I still missing your point?

**Comment by forrestguice** [comment]
@al-caveman Thanks for that link. It mentions an approximate is the midpoint between noon and sunset - I think the [IntervalMidpoint](https://github.com/forrestguice/IntervalMidpoints) addon could be used to schedule alarms for that time.

> Are you saying that Google requires that, if an app supports, say, API 14+, Google rejects it, despite the fact that API 14+ also supports 31+? I.e. Google forces developers to artificially forbid too-old devices from being able to get an app, even if newer devices are still supported?

Something like that. https://apilevels.com/ has a lot of good technical information. Google requires targeting the latest (arguably a good thing from a security standpoint). The issue is the "support libraries" when renamed "AndroidX libraries" increased their minimum to 14. Anyone that wants to target later versions must upgrade those libraries, forcing them to drop support. I don't think Google has explicit policy that forbids old devices, just evolving technical requirements that make it impossible. I doubt many devices still run 14, but now 19 is considered "ancient", and 21 "old".

**Comment by Timo-Fi** [comment]
[al-caveman](https://github.com/al-caveman), I took a closer look at this asr-time:  

So, we have:
H == height of the stick
S == the length of stick's shadow at noon

To get the S we need to specify:
NA == altitude angle of the sun at noon

Then the altitude angle of the sun at the asr can be calculated, I name that angle as TA. The time value at TA would need much more calculation (Suntimes dialog provides that) but TA in itself gives a decent view to the asr. I'm showing that, at least for the photographers eye, the asr moment looks very much different depending on the altitude of the sun at noon, that in turn depends on the day of the year and on the latitude of the particular location. So, as the height of the stick (H) reduces away, we get:

TA = atan(1 / (2 + 1 / tan(NA)))

The below lists shows the altitude TA by 10° steps of NA:

NA° / TA°
10 / 7.4
20 / 11.9
30 / 15.0
40 / 17.4
50 / 19.4
60 / 21.2
70 / 22.9
80 / 24.7
90 / 26.6

E.g. at my position (Helsinki/Finland) we get the following:

Solstice / NA° / TA°
Winter / 6.6 / 5.4
Summer / 53.3 / 20

This means that at the Winter Solstice the ars-moment happens just +/- 65 minutes from the noon. The 20° altitude of the sun at the Summer Solstice, by the conception of many Fins, equals to "bright daylight" (because we never have the high sun).

Surely the definition you have given, results a time vale in between the sunrise and noon and in between the noon and the sunset. However I must say that I can not see what practical use it could serve, as at a single location it is fluctuating over the path of a year, and at same date it looks different from latitude to latitude.  

Just my 2¢

**Comment by Friptick** [comment]
This custom event appears to include both rising and falling phases, whereas the alarm event allows for two specific events, is that correct?

Is that for a reason or is it just an oversight?

Thanks for your work.

**Comment by forrestguice** [comment]
@Friptick sorry, I'm not sure I follow.

These should be handled the same as the rising and setting twilight times.
When setting an alarm there is a selection between rising and setting, but in other places (widget configuration, UI settings), those are displayed as a single combined item. Is this what you're referring to?

**Comment by forrestguice** [comment]
It's also possible you're seeing a bug in the alarm selector. 🤷‍♂️ 
It's the "edge of the UI", so this is where they like to hang out ;)

There should be a way to `v` and `^` each row when adding an alarm to reveal that each item is actually two (rising, setting).

**Comment by Friptick** [comment]
>n other places (widget configuration, UI settings), those are displayed as a single combined item. Is this what you're referring to?

Yep exactly. Seems like an oversight, no? After all it's possible to distinguish between the two with the alarm (which works great!).

I.e., I have an alarm set for 30 degrees in the afternoon _only_. I'd similarly like to see the upcoming event for 15 degrees but in the morning _only_, without seeing the irrelevant one for 15 degrees in the afternoon too.

**Comment by forrestguice** [comment]
> Seems like an oversight, no?

hrm, I'm not sure... these "custom events" are just additional twilights and the main UI and widgets currently display those in pairs. So its consistent with those other built-in events. When it comes to alarms though it makes sense that you can only choose one.

_**It sounds like you want a widget that displays only a single event? e.g. sunrise only.**_
I think it makes sense to add widgets capable of doing that. 👍 

Off the top of my head, I think the 1x1 flippable widget displays one event at a time.
I just tried it though and it seems the flip part is now broken - I'm assuming its a bug introduced in `v0.16.0`.

**Comment by Friptick** [comment]
>these "custom events" are just additional twilights and the main UI and widgets currently display those in pairs. So its consistent with those other built-in events. 

Indeed. Didn't consider this properly. Would be bad UX to change only one of them.

Let me describe how I'd like to use the app (to be clear: it's already *great*, this would just make it perfect!)

I have several daily events that I'm interested in. Let's say they are:

1. Sunrise
2. 20 degrees sun elevation in the afternoon
3. 1 hour before golden hour

Each event happens once per day. So 20 degrees elevation in the *morning* is irrelevant, sun*set* is irrelevant, etc. I'd guess this is the way most users roll (but it's only a guess).

Ideally I'd like to just open the main UI (without resort to widgets) and see *how much time until the next of my events*.

The alarm is just for a particularly important event (i.e. one where I may be sleeping). In my mental model, the events themselves are functionally identical. At any given moment, I'd just like to be able to know when the next one is coming up.

Hope that's useful. Let me know if I should move this to a feature request (NB it's almost there already!). Thanks for your work.

**Comment by forrestguice** [comment]
> Hope that's useful.

Very useful! I'll have to give this some thought.
